### DISCUSSION TITLE
How to extend OpenAM functionality to get data from a DB and enable a service to get this data?

### DISCUSSION THREAD

**Discussion by korah91** [OpenIdentityPlatform/OpenAM#809] [Q&A] [answered]
Hi, I need to extend OpenAM 13.5 functionality in order to get some data from an OracleDB when an application calls certain URL of OpenAM including a valid user token.
I don't understant the docs, I don't have a lot of experience in Java and would love some help from an expert. I need to "enable" some endpoint so that when the app calls this endpoint a script gets activated, gets the data from the DB using the user's information and the application number, and finally saves that into the user's session. All of this so that then, the application can call the /sessions/?action_getProperties to get the data from the DB that has been saved in the session properties.

Thanks for reading this!

**Comment by maximthomas** [comment, marked as answer]
Hi, you can extend OpenAM as described in the [article](https://github.com/OpenIdentityPlatform/OpenAM/wiki/How-to-Customise-OpenAM) in the wiki and add a servlet with the required functionality. However, this approach requires some knowledge of Java and OpenAM's inner architecture.

You can also ask for commercial support [here](https://github.com/OpenIdentityPlatform/.github/wiki/Approved-Vendor-List)

**Reply by korah91** [reply]
Thanks for the reply, Maxim!
Is there any repository with extensions/servlets that people have made?

**Reply by maximthomas** [reply]
I don't know about public repositories with modified OpenAM, but [this](https://docs.oracle.com/cd/E13222_01/wls/docs92/webapp/configureservlet.html) article from Oracle could be a good starting point.

**Comment by vharseko** [comment]
* [Build OpenAM from source](https://github.com/OpenIdentityPlatform/OpenAM/blob/master/README.md#how-to-build)
* [How to make OpenAM log more verbose](https://github.com/OpenIdentityPlatform/OpenAM/wiki/How-to-make-OpenAM-log-more-verbose)
* [How to Customise OpenAM](https://github.com/OpenIdentityPlatform/OpenAM/wiki/How-to-Customise-OpenAM)
* [Extending OpenAM](https://doc.openidentityplatform.org/openam/dev-guide/chap-extending)
* [OpenAM JDBC Authentication Module](https://github.com/OpenIdentityPlatform/OpenAM/tree/a1bc44adb0fc660df7612710b84ca332dfae15ef/openam-authentication/openam-auth-jdbc)
* [How to Configure JDBC Authentication Module](https://www.youtube.com/watch?v=VGgt2Aqw5Dc)
* [OpenAM Sample Custom Authentication Module](https://github.com/OpenIdentityPlatform/OpenAM/tree/master/openam-samples/custom-authentication-module)
* [How to Devleop a Custom Authentication Module](https://github.com/OpenIdentityPlatform/OpenAM/wiki/How-to-Devleop-a-Custom-Authentication-Module)
* [Sample Custom OpenAM DataStore Integration from JDBC](https://github.com/OpenIdentityPlatform/OpenAM/tree/master/openam-core/src/main/java/com/sun/identity/idm/plugins/database)
* [How to Create Custom OpenAM DataStore Integration from MongoDB](https://github.com/OpenIdentityPlatform/OpenAM/wiki/How-to-Create-Custom-OpenAM-DataStore-Integration)
* [Sample Custom OpenAM DataStore Integration from Postgres](https://repo1.maven.org/maven2/org/openidentityplatform/openam/jdbcrepo/15.0.0.1/jdbcrepo-15.0.0.1-sources.jar)
* [How to Use Apache Cassandra as User DataStore in OpenAM](https://github.com/OpenIdentityPlatform/OpenAM/tree/master/openam-cassandra/openam-cassandra-datastore)
